File: src/cli/query/aggregate.ts
import { Command, RangeArgs, ref } from "vipvot";
import type { Document } from "mongodb";
import { printJson, printError, printNdjsonStream, resolvePageSize } from "../../lib/output.ts";
import { getSettings } from "../../lib/config.ts";
import { getMongoClient, closeAllClients } from "../../mongo/client.ts";
import { runAggregate, streamAggregate } from "../../mongo/aggregate.ts";
import { enhanceErrorMessage } from "../../lib/errors.ts";
import { parseJsonArray } from "../../lib/parse-json.ts";
import { resolveConnectionAlias } from "../_globals.ts";

export function buildAggregateCommand(): Command {
  const pipelineFlag = ref<string>("");
  const limit = ref<string>("");
  const stream = ref<boolean>(false);

  const cmd = Command({
    use: "aggregate <database> <collection> [pipeline]",
    short: "Run a read-only aggregation pipeline",
    args: RangeArgs(2, 3),
    run: async (_cmd, args) => {
      const [database, collection, pipelineArg] = args as [string, string, string?];
      try {
        const { client } = await getMongoClient(resolveConnectionAlias());

        const pipeline = await resolvePipeline(pipelineArg, pipelineFlag.value || undefined);

        if (stream.value) {
          const cursor = streamAggregate(client, {
            dbName: database,
            collName: collection,
            pipeline,
          });
          await printNdjsonStream(cursor);
        } else {
          const maxDocs = getSettings().query?.maxDocuments ?? 100;
          const requestedLimit = resolvePageSize({ limit: limit.value || undefined });
          const limitVal = Math.min(requestedLimit, maxDocs);

          const result = await runAggregate(client, {
            dbName: database,
            collName: collection,
            pipeline,
            limit: limitVal,
          });
          printJson({ database, collection, ...result });
        }
      } catch (err) {
        printError(
          err instanceof Error
            ? enhanceErrorMessage(err, { database, collection })
            : "Failed to run aggregation",
        );
      } finally {
        await closeAllClients();
      }
    },
  });

  cmd
    .flags()
    .stringVarP(
      pipelineFlag,
      "pipeline",
      "",
      "",
      "Aggregation pipeline as JSON array (or pipe via stdin)",
    );
  cmd.flags().stringVarP(limit, "limit", "", "", "Max results if pipeline has no $limit stage");
  cmd
    .flags()
    .boolVarP(
      stream,
      "stream",
      "",
      false,
      "Stream results as NDJSON (one JSON object per line, no limit)",
    );

  return cmd;
}

async function resolvePipeline(positionalArg?: string, pipelineFlag?: string): Promise<Document[]> {
  let raw: string;

  if (positionalArg) {
    raw = positionalArg;
  } else if (pipelineFlag) {
    raw = pipelineFlag;
  } else if (!process.stdin.isTTY) {
    raw = await readStdin();
  } else {
    throw new Error(
      "Provide pipeline as argument, --pipeline <json>, or pipe a JSON array via stdin.",
    );
  }

  return parseJsonArray(raw, "pipeline") as Document[];
}

async function readStdin(): Promise<string> {
  const chunks: string[] = [];
  process.stdin.setEncoding("utf8");
  for await (const chunk of process.stdin) {
    chunks.push(chunk as string);
  }
  const result = chunks.join("").trim();
  if (!result) {
    throw new Error(
      "Empty stdin. Provide pipeline as argument, --pipeline <json>, or pipe a JSON array via stdin.",
    );
  }
  return result;
}

File: internal/cli/query/aggregate.go
package query

import (
	"errors"
	"io"
	"maps"
	"os"
	"strings"

	"github.com/spf13/cobra"
	"go.mongodb.org/mongo-driver/v2/bson"

	"github.com/shhac/agent-mongo/internal/cli/shared"
	"github.com/shhac/agent-mongo/internal/ejson"
	"github.com/shhac/agent-mongo/internal/mongo"
	"github.com/shhac/agent-mongo/internal/output"
	"github.com/shhac/agent-mongo/internal/serialize"
)

func registerAggregate(parent *cobra.Command, globals func() *shared.GlobalFlags) {
	var pipelineFlag string
	var limit int
	var stream bool

	cmd := &cobra.Command{
		Use:   "aggregate <database> <collection> [pipeline]",
		Short: "Run a read-only aggregation pipeline",
		Args:  cobra.RangeArgs(2, 3),
		RunE: func(_ *cobra.Command, args []string) error {
			g := globals()
			ref := mongo.Ref{DB: args[0], Collection: args[1]}

			positional := ""
			if len(args) == 3 {
				positional = args[2]
			}
			pipeline, err := resolvePipeline(positional, pipelineFlag)
			if err != nil {
				return err
			}

			maxResults := aggregateLimit(pipeline, limit)
			return shared.WithSessionRef(g, ref, func(ctx shared.SessionCtx) error {
				result, err := ctx.Session.Aggregate(ctx.Ctx, mongo.AggregateOpts{
					Ref:      ref,
					Pipeline: pipeline,
					Limit:    maxResults,
				})
				if err != nil {
					return err
				}
				return printAggregate(result, ref, pipeline, maxResults)
			})
		},
	}

	cmd.Flags().StringVar(&pipelineFlag, "pipeline", "",
		"Aggregation pipeline as JSON array (or pipe via stdin)")
	cmd.Flags().IntVar(&limit, "limit", 0, "Max results if the pipeline has no $limit stage (a pipeline's own $limit is capped at query.maxDocuments)")
	cmd.Flags().BoolVar(&stream, "stream", false, "Deprecated no-op: NDJSON is the default output")
	_ = cmd.Flags().MarkHidden("stream")
	parent.AddCommand(cmd)
}

// aggregateLimit is the cap an aggregation runs under. A pipeline with a $limit
// of its own decides how many results it wants, up to query.maxDocuments like
// every other query; one without gets --limit or the default page size.
func aggregateLimit(pipeline bson.A, flag int) int {
	if mongo.HasLimitStage(pipeline) {
		return shared.MaxDocuments()
	}
	return shared.EffectiveLimit(flag)
}

func resolvePipeline(positional, flag string) (bson.A, error) {
	raw := positional
	if raw == "" {
		raw = flag
	}
	if raw == "" {
		stdin, err := readStdin()
		if err != nil {
			return nil, err
		}
		raw = stdin
	}
	return ejson.ParseArray(raw, "pipeline")
}

func readStdin() (string, error) {
	info, err := os.Stdin.Stat()
	if err != nil || info.Mode()&os.ModeCharDevice != 0 {
		return "", errors.New(
			"Provide pipeline as argument, --pipeline <json>, or pipe a JSON array via stdin.")
	}
	data, err := io.ReadAll(os.Stdin)
	if err != nil {
		return "", err
	}
	trimmed := strings.TrimSpace(string(data))
	if trimmed == "" {
		return "", errors.New(
			"Empty stdin. Provide pipeline as argument, --pipeline <json>, or pipe a JSON array via stdin.")
	}
	return trimmed, nil
}

// printAggregate assembles the aggregation output and its optional echo, split
// from the command so it is testable without a live session.
func printAggregate(result mongo.AggregateResult, ref mongo.Ref, pipeline bson.A, limit int) error {
	meta := output.Meta(map[string]any{
		"database":   ref.DB,
		"collection": ref.Collection,
		"count":      len(result.Documents),
	})
	maps.Copy(meta, output.PaginationMeta(result.HasMore, "", 0))
	var e echo
	// Both orders matter: the sequence of stages, and each stage's fields.
	e.value("pipeline", serialize.OrderedValue(pipeline))
	e.num("limit", limit)
	maps.Copy(meta, e.meta())
	return output.PrintList(result.Documents, meta)
}
